### PR TITLE
Add Cache for CI/CD github action

### PR DESCRIPTION
**Pull request by sishilong**
Add Cache for CI/CD github action. I saw the repeated run after first run is 1 min https://github.com/michelangelo-ai/michelangelo/actions/runs/12471137926

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -16,10 +16,22 @@ jobs:
       - name: Checkout Repository
         uses: actions/checkout@v3
 
+      # Cache Bazel outputs
+      - name: Cache Bazel Outputs
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cache/bazel
+            bazel-out
```
Comment (linzhp) [inline .github/workflows/main.yml]: bazel-out is actually a symlink pointing to `~/.cache/bazel/_bazel_<user>`. If we cache the whole `~/.cache/bazel`, it already incldues bazel-out

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -16,10 +16,22 @@ jobs:
...
+          key: ${{ runner.os }}-bazel-${{ hashFiles('WORKSPACE', '**/BUILD.bazel') }}
```
Comment (linzhp) [inline .github/workflows/main.yml]: Does this mean whenever `WORKSPACE` or any BUILD.bazel files change, the whole `~/.cache/bazel` is nuked?

If that's the case, we can put a static key like "bazel_cache_root" so it always has a warm cache.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -16,10 +16,22 @@ jobs:
       - name: Checkout Repository
         uses: actions/checkout@v3
 
+      # Cache Bazel outputs
+      - name: Cache Bazel Outputs
+        uses: actions/cache@v3
```
Comment (linzhp) [inline .github/workflows/main.yml]: I saw the latest version is `actions/cache@v4`

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -16,10 +16,22 @@ jobs:
...
+      # Cache Bazel outputs
+      - name: Cache Bazel Outputs
+        uses: actions/cache@v4
+        with:
+          path: |
+            ~/.cache/bazel
+          key: bazel_cache_root
+          restore-keys: |
+            bazel_cache_root
+            ${{ runner.os }}-bazel-
```
Comment (linzhp) [inline .github/workflows/main.yml]: we probably don't need the second key?

**Review by linzhp — APPROVED** [review]

File: .github/workflows/main.yml (hunk 4 above)
Comment (sishilong) [reply .github/workflows/main.yml]: I added the second key for fallback. but can remove it now. we can monitor how the ci/cd behavior

**Pull request merged**